### PR TITLE
change(web): drops old IE-related handling code

### LINKED ISSUE #5943: chore(web): remove remaining IE references
**Describe the bug**

There are still a number of references to IE11 and older versions, e.g. mspointerevents, promise polyfill, _IE variable, IEVersion, document.createEvent, etc. These should be removed.

---

**KeymanWeb:**
 - KeymanWeb Version: 15.0

### PR DESCRIPTION
**Pull request by jahorton**
Fixes #5943.
Also addresses some concerns of #3796.  As of commit https://github.com/keymanapp/keyman/pull/6557/commits/474634a90ed9657069ddf8b4518d863b89d649f5,

image

While this _should_ be safe, removing numerous old internal hooks like this could have unforeseen complications.

These changes are 16.0-only; they will not land in 15.0.

This PR seeks to remove a lot of the older parts of KMW that were devoted to IE-wrangling.  We haven't actively tested KMW on IE in a while and we officially dropped support versions ago.  Now, it's time to drop all pretenses of IE support entirely.

That said, we still do need to keep some polyfills around, as older Android devices that we still support launched with old enough versions of Chrome that parts of KMW will instantly break on them unless their users actually update Chrome.  I added documenting comments to the most relevant file:

https://github.com/keymanapp/keyman/blob/326dc4d72054dc51b16b67029a87029b243c4676/common/web/lm-worker/build.sh#L54-L80

This is the LMLayer's worker polyfill set.  Note that some of these polyfills are used explicitly because we don't want `es6-shim` bloat in the `WebWorker` - KeymanWeb itself would use `es6-shim` instead for such cases.  That said, KMW doesn't actually seem to utilize the features being polyfilled... here.

Also, now that we're dropping IE... we can actually just straight-up stop bundling `es6-shim` into KMW.

# User Testing 

## Test Suite 1:  in-app uses

SUITE_IN_APP:

### Testing Environments

GROUP_ANDROID_5_APP:  Android 5.0 / API 21 (any matching device; emulation is fine)
- Do NOT update Chrome if requested.
- If not requested and in an emulated environment, please reset the device to its "original settings" if possible.

GROUP_IOS_12_APP: iOS 12.0 (any matching device; Simulator is fine)

For both, please:
- Have `sil_euro_latin` (the default keyboard) installed with predictive text active.
- Have any other keyboard installed for at least one language.

### Tests

- **TEST_INAPP_BASE**:  Using English(EuroLatin SIL)...
  In portrait orientation, verify OSK is visible and fills the width the bottom of the screen

  1. Verify long-press `q` key works
  2. Verify long-press `p` key works
  3. Verify uppercase layer can be selected via `SHIFT`
  4. Verify number layer can be selected via `123`
  5. Verify long-press `1` key works
  6. Verify long-press `0` key works
  7. Verify backspace, space, and enter keys work
  8. Verify *Khmer* -> *Khmer Angkor* keyboard can be added via Settings menu

- **TEST_INAPP_LONG_PRESS**:  Using English(EuroLatin SIL)...

  Type on the OSK using the following scenarios and verify expected output:
  1. Clicking a suggestion on the suggestion banner - should insert the suggestion
  2. short-press a key and release - should insert the base key
  3. long-press a key, select a long-press key, and release - should insert the long-press key

- **TEST_KEYBOARD_SWAP**:  Use the keyboard's globe key to swap keyboards.
    - If any "Fatal Error" appears or the selected keyboard fails to load properly, _**FAIL**_ this test.

- **TEST_PREDICTIVE_TEXT**:  Using English(EuroLatin SIL)...
    1.  Use the app's "Clear Text" command to erase all text.
    2. The usual default suggestions should appear:  "the", "to", "a".
    3. Type `th`.  All suggestions should begin with these two letters.
    4. Select one of the suggestions.  It should replace the `th` with your selection.

## Test Suite 2:  standalone use on mobile platforms

SUITE_BROWSER:

### Testing Environments

GROUP_ANDROID_6_CHROME:  Chrome on Android **6.0 / API 23** (any matching device; emulation is fine)
- Note:  _**not**_ 5.0.  We do extra things in the app to support it that are no longer done for standalone-Web.
    - In particular, note the comments below regarding `for ... of` and `Map`.
GROUP_IOS_12_SAFARI: Safari on iOS 12.0 (any matching device; Simulator is fine)

For both, please:
- Have `sil_euro_latin` (the default keyboard) installed with predictive text active.
- Have any other keyboard installed for at least one language.

### Tests

- **TEST_INAPP_BASE**:  Use the testing page "Test unminified Keymanweb".
  In portrait orientation, verify OSK is visible and fills the width the bottom of the screen

  1. Verify long-press `q` key works
  2. Verify long-press `p` key works
  3. Verify uppercase layer can be selected via `SHIFT`
  4. Verify number layer can be selected via `123`
  5. Verify long-press `1` key works
  6. Verify long-press `0` key works
  7. Verify backspace, space, and enter keys work
  8. Verify *Khmer* -> *Khmer Angkor* keyboard can be added via Settings menu

- **TEST_LOREM_IPSUM**:  Use the testing page "Tests performance of touch-alias elements with large amounts of text" - it should be the last one listed on the testing index page.

    - Touch and drag the block of text, then tap somewhere within the text.  The caret should be placed near the location of the tap after a very small delay.  A bit of lagginess is fine, but it should still feel reasonably responsive.
    - Repeat this at least three times.
    - If the delay ever goes longer than one second, **_FAIL_** this test.
    - If the caret is not placed properly at any point, **_FAIL_**  this test.
    - After the final caret placement, type `lorem ipsum`.  Text entry should feel responsive and no laggier than the other interactions required by this test.

### CODE REVIEW

**Comment by jahorton** [comment]
Ooh, a "fun" note:  when testing on an un-updated Android API 21 (5.0) image, most KMW testing pages will break!

To fix this, replace `web/testing/commonHeader.js`'s version of this function with this one:

```
  function errToString(err) {
    // Painful?  Kinda.  But needed on un-updated Android API 21!
    if(Array.isArray(err)) {
      var result = '';
      for(var i = 0; i < err.length; i++) {
        var e = err[i];
        if(e.error instanceof Error) {
          result += e.error.message + '\n';
        } else {
          result += JSON.stringify(e) + '\n';
        }
      }
      return result;
    }
    if(err instanceof Error) {
      return err.message;
    }
    return JSON.stringify(err);
  }
```

Neither `let` nor `for...of` is supported in the unupdated version of its browser!

**Comment by jahorton** [comment]
For the previous commit (https://github.com/keymanapp/keyman/pull/6557/commits/f14e7f1a38d91251afd4d38ec52f0aec8215fa82), I dummied out the `PromiseStore` `Map` dependence issue noted in this comment on the base issue: https://github.com/keymanapp/keyman/issues/5943#issuecomment-1106663719

After doing so, the standard unminified test page appeared fully functional on an emulated API 21 device running Chrome for Android 37.

**Comment by jahorton** [comment]
Note tagged PRs - there are a few comments noted here that motivate use of arrow functions, which had always been an issue for IE.  Adjustments will be possible once `beta` is merged into `master` near the end of sprint.

**Comment by jahorton** [comment]
Extra note:  make absolutely sure that IE has been removed from all test configs for Web - such as in the "manual" (local) unit-testing configuration on Windows.

File: common/web/lm-worker/build.sh [hunk 1]
```diff
@@ -44,56 +44,46 @@ WORKER_TEST_BUNDLE_TARGET_FILENAME=$WORKER_OUTPUT/index.wrapped-for-bundle.ts
...
+  cat "src/polyfills/symbol-es6.min.js" || die # Needed for Android / Chromium browser pre-43.
 
-  # Needed to 'support' String.normalize within iOS 9. For our limited use case
-  # thereof; is definitely NOT a general polyfill. (The file size on a true one
-  # would be quite high.)
+  # Needed to 'support' String.normalize within iOS 9.
+  # For our limited use case thereof; is definitely NOT a general polyfill.
+  # (The file size on a true one would be quite high.)
+  # TBD:  should we drop this, since we only support iOS 12+ for the mobile app?
   cat "src/polyfills/string.normalize.js" || die
```
Comment (jahorton) [inline common/web/lm-worker/build.sh:81]: On revisiting the PR and merging related changes, I added the `# TBD` there.  Do we keep the iOS 9 piece in, or should we go ahead and drop it?

Comment (mcdurdin) [reply common/web/lm-worker/build.sh:81]: Drop it!

Comment (mcdurdin) [reply common/web/lm-worker/build.sh:81]: (Assuming our Chrome support is ok)

**Comment by bharanidharanj** [comment]
## SUITE_IN_APP: 

### GROUP_IOS_12_APP: iOS 12.0 (any matching device; Simulator is fine)

* **TEST_INAPP_BASE (PASSED):** Tested Keyman 15.0.260-alpha build in iPhone 13 Pro Simulator, as per the instructions it is working as expected. 
* **TEST_INAPP_LONG_PRESS (PASSED):**  Tested Keyman 15.0.260-alpha build in iPhone 13 Pro Simulator, as per the instructions it is working as expected. 
* **TEST_KEYBOARD_SWAP (PASSED):**  Tested Keyman 15.0.260-alpha build in iPhone 13 Pro Simulator, as per the instructions it is working as expected. 
* **TEST_PREDICTIVE_TEXT (PASSED):**  Tested Keyman 15.0.260-alpha build in iPhone 13 Pro Simulator, as per the instructions it is working as expected.

**Comment by bharanidharanj** [comment]
## SUITE_BROWSER: 

### GROUP_ANDROID_6_CHROME: Chrome on Android **6.0 / API 23** (any matching device; emulation is fine)

* **TEST_INAPP_BASE (PASSED):** notes
* **TEST_LOREM_IPSUM (SKIPPED):** notes

**Comment by jahorton** [comment]
Weird that you're getting a 404 when following that link; it works perfectly fine for me right now.  Maybe something went funny with the server when you were testing?

BTW, I believe that in this scenario, that'd be a "BLOCKED", not a "FAILED" - you didn't even get to run the tests, after all.

SUITE_BROWSER 

GROUP_ANDROID_6_CHROME TEST_LOREM_IPSUM OPEN
GROUP_IOS_12_SAFARI TEST_LOREM_IPSUM OPEN

**Comment by bharanidharanj** [comment]
> Weird that you're getting a 404 when following that link; it works perfectly fine for me right now. Maybe something went funny with the server when you were testing?
> 
> BTW, I believe that in this scenario, that'd be a "BLOCKED", not a "FAILED" - you didn't even get to run the tests, after all.
> 
> SUITE_BROWSER
> 
> GROUP_ANDROID_6_CHROME TEST_LOREM_IPSUM OPEN
> 
> GROUP_IOS_12_SAFARI TEST_LOREM_IPSUM OPEN

@jahorton Ah..Yes. It is "BLOCKED" and not "FAILED". Thanks Josh, for the clarification. 👍

**Comment by mcdurdin** [comment]
> * Well... unless we have Android itself include `es6-shim` and insert it into the keyboard host page if the Chrome WebView isn't sufficiently updated.  (Won't fix unembedded Web, but it'll at least prevent the app from breaking.)

This is my preference: Keyman Engine for Web should not include *any* polyfills. Any app that uses Keyman Engine for Web (whether that be a native device app like Keyman for Android, or a website like keymanweb.com) can decide which versions of browsers it wants to support and provide the necessary polyfills. This is similar to our strategy for supporting error reporting -- it's up to the app, not Keyman Engine for Web, to make these choices.

(The worker may be an issue, admittedly -- any ideas?)

What we should do, to be nice JS citizens, is provide guidance in the Keyman Engine for Web documentation on which features may require polyfills.

Note: if we want keymanweb.com to support older browsers, we can include es6-shim.js there. Analysis of current device users may help on this.

File: common/web/lm-worker/build.sh [hunk 2]
```diff
@@ -44,56 +44,40 @@ WORKER_TEST_BUNDLE_TARGET_FILENAME=$WORKER_OUTPUT/index.wrapped-for-bundle.ts
 wrap-worker-code ( ) {
   name="$1"
   js="$2"
-  # Reference: https://stackoverflow.com/a/59046041
-  # Use of this annotation allows us to actually build against the file, rather
-  # than hackily insert it later when/where needed.
   echo "// @ts-nocheck"
   echo "// Autogenerated code. Do not modify!"
-  echo "// --START:LMLayerWorkerCode--"
```
Comment (mcdurdin) [inline common/web/lm-worker/build.sh:52]: Is there a reason you've dropped these?

File: web/source/dom/touchAliasElement.ts [hunk 3]
```diff
@@ -542,21 +539,12 @@ namespace com.keyman.dom {
 
       // snapOrder - 'snaps' the touch location in a manner corresponding to the 'ltr' vs 'rtl' orientation.
       // Think of it as performing a floor() function, but the floor depends on the origin's direction.
-      let snapOrder: (x: number, y: number) => boolean;
+      const isRTL = (this as unknown as HTMLElement).dir == 'rtl';
+      const snapOrder = isRTL ? (a, b) => a < b : (a, b) => a > b;
```
[the remark this answers is not in the document]
Comment (mcdurdin) [reply web/source/dom/touchAliasElement.ts:543]: > Except... they're actually transpiling now?

Yeah, I noticed that they've been transpiling properly for a while. Perhaps we had the wrong output version set at some point?

File: web/source/kmwdebug.js [hunk 4]
```diff
@@ -20,8 +20,6 @@ if(!window['keyman']['initialized']) {
     {
       var dhost = document.createElement('DIV');
       dhost.style.display = 'block';
-      if(!keymanweb._IE)
-        dhost.style.position = 'fixed';
```
Comment (mcdurdin) [inline web/source/kmwdebug.js:24]: Is this right? You've deleted code that's `!_IE`?

File: web/source/kmwdebug.js [hunk 5]
```diff
@@ -32,8 +30,6 @@ if(!window['keyman']['initialized']) {
       
       keymanweb._DivDebug = document.createElement('DIV');
       keymanweb._DivDebug.style.fontFamily = 'Lucida Console,Courier New,courier';
-      if(!keymanweb._IE)
-        keymanweb._DivDebug.style.position = 'absolute';
```
Comment (mcdurdin) [inline web/source/kmwdebug.js:36]: Is this right? You've deleted code that's `!_IE`?

File: web/source/kmwdebug.js [hunk 6]
```diff
@@ -45,8 +41,6 @@ if(!window['keyman']['initialized']) {
       dhost.appendChild(keymanweb._DivDebug);
       
       var _dd = document.createElement('DIV');
-      if(!keymanweb._IE)
-        _dd.style.position = 'absolute';
```
Comment (mcdurdin) [inline web/source/kmwdebug.js:49]: Is this right? You've deleted code that's `!_IE`?

File: web/source/kmwuitoggle.ts [hunk 7]
```diff
@@ -584,7 +584,7 @@ if(!window['keyman']['ui']['name']) {
       "#KeymanWeb_KbdList {"+
         "display: block;"+
         "position: absolute;"+
-        "width: auto;"+ ((util['getIEVersion']() < 8) ? "200px;" : "auto;")+
```
Comment (mcdurdin) [inline web/source/kmwuitoggle.ts:587]: Huh. Was this just broken CSS previously?

File: web/source/kmwutils.ts [hunk 8]
```diff
@@ -304,17 +304,12 @@ namespace com.keyman {
...
-        e.style.KhtmlUserSelect="none";
-        e.style.UserSelect="none";
-        e.style.WebkitUserSelect="none";
-      }
+      const e = document.createElement<E>(nodeName);
+
+      e.style.MozUserSelect="none";
+      e.style.KhtmlUserSelect="none";
+      e.style.UserSelect="none";
+      e.style.WebkitUserSelect="none";
```
Comment (mcdurdin) [inline web/source/kmwutils.ts:312]: Three of these are probably also unnecessary now but I haven't checked version support for *UserSelect?

**Review by mcdurdin — COMMENTED** [review]
First-pass review :grin:

**Comment by jahorton** [comment]
> (The worker may be an issue, admittedly -- any ideas?)

Always has been.  That'd take some time and investment to work around, I think.

------

For inlined scripts:  https://stackoverflow.com/a/34854903

For file-based scripts:  it sounds like something AJAX based would be required.

**Comment by bharanidharanj** [comment]
## SUITE_BROWSER: 

### GROUP_ANDROID_6_CHROME: Chrome on Android **6.0 / API 23** (any matching device; emulation is fine)

* **TEST_LOREM_IPSUM (BLOCKED):** Not able to test this issue, since I am getting 404 error message.

**Comment by bharanidharanj** [comment]
## SUITE_BROWSER: 

### GROUP_IOS_12_SAFARI: Safari on iOS 12.0 (any matching device; Simulator is fine)

* **TEST_LOREM_IPSUM (FAILED):** Not able to test this issue, since I am getting 404 error message.

**Comment by mcdurdin** [comment]
> * **TEST_LOREM_IPSUM (BLOCKED):** Not able to test this issue, since I am getting 404 error message. 
> 
>  
> 

@bharanidharanj  I am able to replicate this error page. We will fix that in an upcoming update to the test pages. However, I am not sure why you need to click that link in order to run the test? You should be able to navigate from the User Test Results artifacts section:

image

to:

image

and from there, to:

image

without ever needing to click the "Return to main index" link? (If you accidentally go to the wrong page, use the back button in your browser)

@keymanapp-test-bot retest TEST_LOREM_IPSUM

File: web/source/kmwdebug.js (hunk 4 above)
Comment (jahorton) [reply web/source/kmwdebug.js:24]: Oops.  I probably played it a little too fast and loose on this particular file; I've never once actually used its code.  It's just kind of... lingered on, so far as I know.

File: web/source/kmwutils.ts (hunk 8 above)
Comment (jahorton) [reply web/source/kmwutils.ts:312]: https://developer.mozilla.org/en-US/docs/Web/CSS/user-select

The fact that the article still references all of them and their differences is something I find concerning.

File: common/web/lm-worker/build.sh [hunk 9]
```diff
@@ -52,48 +52,36 @@ wrap-worker-code ( ) {
   echo "// --START:LMLayerWorkerCode--"
   printf "function %s () {\n" "${name}"
 
-  # TODO: I don't know if this is still true, or if we can add the polyfills to
-  # the project directly now?
```
Comment (jahorton) [inline common/web/lm-worker/build.sh:56]: Hmm.  We could test this, I suppose.  It's possible the declaration emitter got enhanced in the newer TS version we're using now.

It definitely fell over quickly when I tried directly linking against it with `allowJS=true` in the config before we were using incremental builds.

**Comment by bharanidharanj** [comment]
> 



> > * **TEST_LOREM_IPSUM (BLOCKED):** Not able to test this issue, since I am getting 404 error message.
> > 
> > 
> 
> @bharanidharanj I am able to replicate this error page. We will fix that in an upcoming update to the test pages. However, I am not sure why you need to click that link in order to run the test? You should be able to navigate from the User Test Results artifacts section:
> 
> image
> 
> to:
> 
> image
> 
> and from there, to:
> 
> image
> 
> without ever needing to click the "Return to main index" link? (If you accidentally go to the wrong page, use the back button in your browser)
> 
> @keymanapp-test-bot retest TEST_LOREM_IPSUM

@mcdurdin  I thought by clicking on the 'Return to main index' link would show the corresponding test page. Anyhow, thanks for your suggestion. I will retest it and post my comment. 👍

**Comment by bharanidharanj** [comment]
## SUITE_BROWSER: 

### GROUP_IOS_12_SAFARI: Safari on iOS 12.0 (any matching device; Simulator is fine)

* **TEST_LOREM_IPSUM (PASSED):** Tested in Safari Browser on iOS 12.4 / iPhone 8 Simulator and it seems to be working as expected. ie., Able to see the caret appaers near the location of the tap. After the final caret placement, I am able to type 'lorem ipsum' without any lagging.

**Comment by bharanidharanj** [comment]
## SUITE_BROWSER: 

### GROUP_ANDROID_6_CHROME: Chrome on Android **6.0 / API 23** (any matching device; emulation is fine)

* **TEST_LOREM_IPSUM (PASSED):** Tested in Chrome (version 68.0) on Android 6.0 / API 23 emulator as per the instructions and it seems to be working as expected.

File: common/web/lm-worker/build.sh [hunk 10]
```diff
@@ -52,48 +52,36 @@ wrap-worker-code ( ) {
...
-  # would be quite high.)
-  cat "src/polyfills/string.normalize.js" || die
+  cat "src/polyfills/symbol-es6.min.js" || die # Needed for Android / Chromium browser pre-43.
 
   echo ""
 
-  cat "${js}" || die
+  cat "${js}"
   printf "\n}\n"
-  echo "// --END:LMLayerWorkerCode--"
```
Comment (mcdurdin) [inline common/web/lm-worker/build.sh:96]: This line needs to be restored

File: common/web/lm-worker/build.sh [hunk 11]
```diff
@@ -52,48 +52,36 @@ wrap-worker-code ( ) {
...
-  # Needed to 'support' String.normalize within iOS 9. For our limited use case
-  # thereof; is definitely NOT a general polyfill. (The file size on a true one
-  # would be quite high.)
-  cat "src/polyfills/string.normalize.js" || die
+  cat "src/polyfills/symbol-es6.min.js" || die # Needed for Android / Chromium browser pre-43.
 
   echo ""
 
-  cat "${js}" || die
+  cat "${js}"
```
Comment (mcdurdin) [inline common/web/lm-worker/build.sh:83]: Why not having `|| die` here?

**Review by mcdurdin — APPROVED** [review]
Any change to polyfill strategy (moving responsibility to the consumer of Keyman Engine for Web) belongs in a separate PR, so this LGTM now (two minor comments pending)

**Comment by keyman-server** [comment]
Changes in this pull request will be available for download in [Keyman version 16.0.19-alpha](https://keyman.com/downloads/releases/alpha/16.0.19)

**Pull request merged**